pkg/broker/awsbroker.go

```diff
@@ -241,127 +211,130 @@ func (b *AwsBroker) ValidateBrokerAPIVersion(version string) error {
-				free := k.(bool)
-				plan.Free = &free
-			} else if i.(string) == "metadata" {
-				metadata := make(map[string]interface{})
-				for i2, k2 := range k.(map[interface{}]interface{}) {
-					metadata[i2.(string)] = k2
-				}
-				plan.Metadata = metadata
-			} else if i.(string) == "parameters" {
-				propsForCreate := make(map[string]interface{})
-				requiredForCreate := make([]string, 0)
-				propsForUpdate := make(map[string]interface{})
-				requiredForUpdate := make([]string, 0)
-				for _, param := range k.([]interface{}) {
-					var name string
-					var required, updatable bool
-					pvals := make(map[string]interface{})
-					for pk, pv := range param.(map[interface{}]interface{}) {
-						switch pk {
-						case "name":
-							name = pv.(string)
-						case "required":
-							required = pv.(bool)
-						case "type":
-							switch pv {
-							case "enum":
-								pvals[pk.(string)] = "string"
-							case "int":
-								pvals[pk.(string)] = "integer"
-							default:
-								pvals[pk.(string)] = pv
-							}
-						case "updatable":
-							updatable = pv.(bool)
-						default:
-							pvals[pk.(string)] = pv
-						}
-					}
-					propsForCreate[name] = pvals
-					if required {
-						requiredForCreate = append(requiredForCreate, name)
-					}
-					if updatable {
-						propsForUpdate[name] = pvals
-						if required {
-							requiredForUpdate = append(requiredForUpdate, name)
-						}
-					}
-				}
-				plan.Schemas = &osb.Schemas{
-					ServiceInstance: &osb.ServiceInstanceSchema{
-						Create: &osb.InputParametersSchema{
-							Parameters: map[string]interface{}{
-								"type":       "object",
-								"properties": propsForCreate,
-								"$schema":    "http://json-schema.org/draft-06/schema#",
-							},
-						},
-					},
+	params := cfnParamsToOsb(sd)
+	for k, p := range sd.Metadata.Spec.ServicePlans {
+		planid := uuid.NewV5(db.Accountuuid, "service__"+sd.Metadata.Spec.Name+"__plan__"+k).String()
+		plan := osb.Plan{
+			ID:          planid,
+			Name:        k,
+			Description: p.Description,
+			Free:        &f,
+			Bindable:    &t,
+			Metadata: map[string]interface{}{
+				"cost":            p.Cost,
+				"displayName":     p.DisplayName,
+				"longDescription": p.LongDescription,
+			},
+			Schemas: &osb.Schemas{ServiceInstance: &osb.ServiceInstanceSchema{}},
+		}
+		propsForCreate := make(map[string]interface{})
+		for nk, nv := range nonCfnParamDefs {
+			propsForCreate[nk] = nv
+		}
+		propsForUpdate := make(map[string]interface{})
+		requiredForCreate := make([]string, 0)
+		requiredForUpdate := make([]string, 0)
+		prescribed := make(map[string]string)
+		var openshiftFormCreate []OpenshiftFormDefinition
+		var openshiftFormUpdate []OpenshiftFormDefinition
+		for paramName, paramValue := range params {
+			include := true
+			for planParam, planValue := range p.ParameterValues {
+				if planParam == paramName {
+					include = false
+					prescribed[planParam] = planValue
```

Comment:
Is it correct that there's no way for a plan to change the default parameter value without prescribing it? For example, if I have an RDS template with `db.t2.small` as the default `DBInstanceClass`, I can't create a `production` plan that changes that to `db.r4.large` _and_ allows it to be modified? Not a huge deal, but I'm guessing I'll see a lot more use of the `custom` plan then.

Reply:
that is correct. Currently I'm only setting things like enabling encryption (that make more sense to be enforced/prescribed than instance type) with plan overrides. Would be nice down the line to have a way to change defaults, but still allow the provisioner to choose.

Reply:
oops, I missed that we could do that already with the -spec.yaml implementation. Will add the functionality